pkg/agent/controller/trafficcontrol/controller.go

```diff
@@ -0,0 +1,1009 @@
+					return err
+				}
+			}
+		}
+	}
+
+	// Get name of the return port.
+	targetPortName = c.getPortName(&tc.Spec.TargetPort)
+	// If the name is different from the cached name in the TrafficControl state, it could be caused by the target port
+	// update of the TrafficControl or the creation of the TrafficControl.
+	if targetPortName != tcState.targetPortName {
+		if tcState.targetPortName != "" {
+			// If the stale return port name cached in TrafficControl state is not empty, release the stale target port
+			// from the TrafficControl.
+			_, _, err = c.releaseTrafficControlPort(tcState.targetPortName, tcName)
+			if err != nil {
+				return err
+			}
+			tcUpdated = true
+		}
+	}
+	// Get or create the target port.
+	targetOFPort, _, err := c.getOrCreateTrafficControlPort(&tc.Spec.TargetPort, targetPortName, tcName)
+	if err != nil {
+		return err
+	}
+
+	// Update the TrafficControl state.
+	tcState.targetPortName = targetPortName
+	tcState.returnPortName = returnPortName
+	if tcState.action != tc.Spec.Action {
+		tcState.action = tc.Spec.Action
+		tcUpdated = true
+	}
+	if tcState.direction != tc.Spec.Direction {
+		tcState.direction = tc.Spec.Direction
+		tcUpdated = true
+	}
+
+	// Get the list of Pods applying to the TrafficControl.
+	var podObjects []*v1.Pod
+	if podObjects, err = c.filterPods(&tc.Spec.AppliedTo); err != nil {
+		return err
+	}
+
+	// Reserve the set of Pods in TrafficControl state as stale, and reinitialize the set in TrafficControl state.
+	stalePods := tcState.pods
+	tcState.pods = sets.NewString()
+	// Reserve the set of OF ports in TrafficControl state as old, and reinitialize the set in TrafficControl state.
+	oldOfPorts := tcState.ofPorts
+	tcState.ofPorts = sets.NewInt32()
+	// Iterate the list of Pods applying to the TrafficControl.
+	for _, pod := range podObjects {
+		podNN := k8s.NamespacedName(pod.Namespace, pod.Name)
+		// Insert the Pod to the reinitialized set in TrafficControl state and remove the Pod from the stale set.
+		tcState.pods.Insert(podNN)
+		stalePods.Delete(podNN)
+
+		// If the TrafficControl is not the effective TrafficControl for the Pod, do nothing.
+		if !c.bindPodToTrafficControl(podNN, tcName) {
+			continue
+		}
+
+		// If the TrafficControl is the effective TrafficControl for the Pod, insert the port to the new set in
+		// TrafficControl state.
+		podInterfaces := c.interfaceStore.GetContainerInterfacesByPod(pod.Name, pod.Namespace)
+		if len(podInterfaces) == 0 {
+			klog.InfoS("Interfaces of Pod not found", "Pod", klog.KObj(pod))
+			continue
+		}
+		tcState.ofPorts.Insert(podInterfaces[0].OFPort)
+	}
+
+	// If the target port, direction and action of the TrafficControl is updated, the mark flows should be reinstalled.
+	// If the newly generated ofPort set is different from the stale ofPort set, the flows should be reinstalled, the mark
+	// flows should be also reinstalled.
+	if tcUpdated || len(utilsets.SymmetricDifferenceInt32(tcState.ofPorts, oldOfPorts)) != 0 {
```

Comment:
It should just call `Equal` if you just want to check if they are equal, which is more efficient and readable.